rfcs/0000-central-station.md

```diff
@@ -0,0 +1,507 @@
+
+### Current description
+
+Busloc primarily keeps track of the whereabouts of MBTA buses and publishes a few feeds related to
+them. Busloc runs on the MBTA network because one of the sources of its data, the TransitMaster
+DB, is only accessible from there.
+
+The majority of what Busloc deals with is "normal" bus data. These are buses logged on and running
+their expected, planned routes. Busloc gets this data from TransitMaster, via its hacky internal
+spelunking of the TransitMaster DB, and via Samsara, a cloud service that has more frequent GPS
+updates. Busloc combines these feeds, taking the latest lat/long from either source for every
+vehicle. These logged-on vehicles have their GTFS trip directly in the data, since TransitMaster
+and GTFS both get their source data from HASTUS. The Samsara data doesn't _have_ to be queried
+from the MBTA network, but is done to have all the data in one place and for simplicity merging
+with the TransitMaster data. This data is published as VehiclePositions, which is consumed by
+Swiftly to generate TripUpdates. The data is GTFS-rt-esque, but includes block and run data, which
+Swiftly takes advantage of.
+
+In addition to the planned service, TransitMaster also has "overload" trips, which is when a bus
+is assigned to run a route that it wasn't planned for. These create "ADDED" trips, based on a
+scheduled trip but with an offset at every location. Since Swiftly cannot produce predictions for
+non-GTFS trips (is this still true?), Busloc marks these with a special "unassigned" status,
+preventing Swiftly from incorrectly matching them to a GTFS trip. TransitMaster also generates
+predictions (of unknown quality, or perhaps simply the arrival/departure values from the
+underlying trip + offset), and these are published in a TripUpdates file which concentrate
+consumes directly.
+
+Conversely, "block waivers", when trips are canceled, are also handled by Busloc and also bypass
+Swiftly. Busloc receives the list of waivers via TransitMaster and publishes them as a special
+TripUpdates file, marking their stops as "SKIPPED". Swiftly is meanwhile publishing predictions
+for the canceled trips because the VehiclePositions are in the feed, and Swiftly can't handle
+waivers. So concentrate merges the Swiftly TripUpdates and the special Busloc TripUpdates feed and
+cancels the trips there. (Is that correct?)
+
+Lastly, Busloc has another Samsara feed from Yankee, which includes the VehiclePositions of
+shuttle buses. Those are output more or less as-is, and we don't try to make predictions for them.
+(Do these end up in concentrate? I didn't see them. I can't tell where
+VehiclePositions_enhanced_shuttles.json goes.)
+
+In addition to the standard VehiclePositions and TripUpdates data, Busloc publishes operator data
+(e.g. the name of the person driving the bus) from TransitMaster, for Skate to consume. I believe
+this data is passed along in one of the TripUpdates files (?), and stripped out later in the
```

Comment:
Operator info comes through in `VehiclePositions_enhanced.json`.

Reply:
Operator data is also made available in the Swiftly dashboard. Swiftly Playback is a very useful tool, and some Bus Operations folks use it.